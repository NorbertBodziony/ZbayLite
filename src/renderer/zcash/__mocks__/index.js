import Zcash from '../client'

const requestManager = {
  z_importviewingkey: jest.fn(async () => null),
  z_listreceivedbyaddress: jest.fn(async () => null),
  z_importkey: jest.fn(async () => null),
  z_exportkey: jest.fn(async (address) => `${address}-spending-key`),
  z_getbalance: jest.fn(async () => null),
  z_listunspent: jest.fn(async () => [0]),
  z_sendmany: jest.fn(async () => null),
  z_getoperationstatus: jest.fn(async () => []),
  z_getnewaddress: jest.fn(async (type) => `${type}-private-address`),
  getnewaddress: jest.fn(async () => 'transparent-address'),
  dumpprivkey: jest.fn(async (address) => `${address}-private-key`),
  importprivkey: jest.fn(async (address) => `${address}-private-key`)
}

const client = new Zcash({ requestManager: jest.fn(() => requestManager) })

export const mock = {
  requestManager,
  client
}

export const getClient = jest.fn(() => client)

export default {
  getClient
}
