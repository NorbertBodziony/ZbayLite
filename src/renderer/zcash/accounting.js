import BigNumber from 'bignumber.js'

export default zcashClient => {
  const balance = async address => {
    const balance = await zcashClient.request.z_getbalance(address)
    return new BigNumber(balance)
  }
  const freeUtxos = async address => {
    const utxos = await zcashClient.request.z_listunspent(1, 9999999, false, [
      address
    ])
    return utxos
  }

  return {
    balance,
    freeUtxos
  }
}
