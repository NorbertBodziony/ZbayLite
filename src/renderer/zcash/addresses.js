export default (zcashClient) => {
  const create = async (type = 'sapling') => zcashClient.request.z_getnewaddress(type)
  const createTransparent = async () => zcashClient.request.getnewaddress()
  return {
    create,
    createTransparent
  }
}
