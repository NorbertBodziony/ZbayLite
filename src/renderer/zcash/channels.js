export default {
  general: {
    testnet: {
      name: 'general',
      private: true,
      address:
        'ztestsapling1dfjv308amnk40s89trkvz646ne69553us0g858mmpgsw540efgftn4tf25gts2cttg3jgk9y8lx',
      description:
        'This is a general channel available to all users of Zbay by default.',
      keys: {
        ivk:
          'zivktestsapling1algnz2x84pqcnfdxrlntw73wpuqm3v568cepf5tuctyusm9javpqyjyzqy'
      }
    },
    mainnet: {
      name: 'general',
      private: true,
      address:
        'zs1qm5kdw592l9x704xqtjln3msyn7gqnfat2uxf05yky2u7xzmsuu5ekel3tf8v4ulqnux6u3f4ac',
      description:
        'This is a general channel available to all users of Zbay by default.',
      keys: {
        ivk: 'zivks1y3csr6f8uga948gepdpgq86mxg4dn0a6jx9qnvz07t6jqf53j5rslc9mu3'
      }
    }
  },
  registeredUsers: {
    testnet: {
      name: 'RegisteredUsers',
      private: true,
      address:
        'ztestsapling1ggtntm4dgfc0remdxzex0pz722swjtqxc7wxj4eeqksmups04dkhjvcj5vcklmud2y3yvgj302m',
      description: 'Store data about users.',
      keys: {
        ivk:
          'zivktestsapling1k93jxuecldthr9myy9d0svsy5dz57xvtcw7hyh9wj9xqy4ehfsqszlhdce'
      }
    },
    mainnet: {
      name: 'RegisteredUsers',
      private: true,
      address:
        'zs1n4xfm0rsmju0fyevqwvgu5jve54v6etqv0sa7djv7kjfpwjpv7fsq54nuykax4y820ggy9g74y6',
      description: 'Store data about users.',
      keys: {
        ivk: 'zivks1s4cw25evj50cagzmx8ng2jrsjvefsmpm37znwsdve6cp3fv5eczskzll2h'
      }
    }
  },
  channelOfChannels: {
    testnet: {
      name: 'channelOfChannels',
      private: true,
      address:
        'ztestsapling1rwqrk7uclurs5dqm70g4zf0rh9dmjf60qqtw2n68pa37ycs2zccglcymjg0nnmus70m3wkdyxx0',
      description: 'Contains list of public channels',
      keys: {
        ivk:
          'zivktestsapling194t7fk8ns0v2u4jysmrmwyh2ymq52t5378ey8jym23d2dpy3d5pssyk97t'
      }
    },
    mainnet: {
      name: 'channelOfChannels',
      private: true,
      address:
        'zs1k2kqrssynect92hpaef8ewmgwersrz5apc2gn8eyy9g0pk5gu2a6twdxaytfprkms56qcpafg72',
      description: 'Contains list of public channels',
      keys: {
        ivk: 'zivks14g72xlh44vsy3nk0d7y83q39hs5jkelp0de8l6a8cdkecxwfa5pquav8as'
      }
    }
  },
  // Create new accounts for donations
  zbay: {
    testnet: {
      name: 'donation',
      address:
        'ztestsapling10tfwmgy9jjzvhxxnlfqcakjpaddn8yq9qhwammfmxgf8ln2vkdafr8ut80kplerjer4nktmys07',
      publicKey:
        '03c5c6d4199607a69f4586d00e06432e4e9370d612ce7e075b0aa3240533bb24c4'
    },
    mainnet: {
      name: 'donation',
      address:
        'zs15myy2d3sgn90p0rvcjlamaczqzglsdnqua6a3mz72fy5fhnke9txpg250kmzzndlu096s3d70j0',
      publicKey:
        '03c5c6d4199607a69f4586d00e06432e4e9370d612ce7e075b0aa3240533bb24c4'
    }
  }
}
