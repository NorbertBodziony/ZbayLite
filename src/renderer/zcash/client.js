import native from '../../../native/index.node'
export default class RPC {
  constructor (url = 'https://lightwalletd.zecwallet.co:1443') {
    const result = native.litelib_initialize_existing(url)
    console.log(`Intialization: ${result}`)
  }
  sync = async () => {
    return native.litelib_execute('sync', '')
  }
  rescan = async () => {
    return native.litelib_execute('rescan', '')
  }
  syncStatus = async () => {
    return JSON.parse(native.litelib_execute('syncstatus', ''))
  }
  info = async () => {
    return JSON.parse(native.litelib_execute('info', ''))
  }
  balance = async () => {
    return JSON.parse(native.litelib_execute('balance', ''))
  }
  notes = async () => {
    return JSON.parse(native.litelib_execute('notes', ''))
  }
  sendTransaction = async payload => {
    // TODO add validation of payload
    return native.litelib_execute('send', JSON.stringify(payload))
  }
  list = async (includeMemoHex = 'yes') => {
    return JSON.parse(native.litelib_execute('list', `${includeMemoHex}`))
  }
  height = async () => {
    return JSON.parse(native.litelib_execute('height', '')).height
  }
  quit = async () => {
    return native.litelib_execute('quit', '')
  }
  save = async () => {
    return JSON.parse(native.litelib_execute('save', ''))
  }
  addresses = async () => {
    return JSON.parse(native.litelib_execute('addresses', ''))
  }
  getPrivKey = async address => {
    return JSON.parse(native.litelib_execute('export', address))[0].private_key
  }
  getNewShieldedAdress = async () => {
    return JSON.parse(native.litelib_execute('new', 'z'))[0]
  }
  getNewTransparentAdress = async () => {
    return JSON.parse(native.litelib_execute('new', 't'))[0]
  }
}
