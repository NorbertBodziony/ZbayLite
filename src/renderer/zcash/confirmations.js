export default (zcashClient) => {
  const getResult = async (txId) => {
    const res = await zcashClient.request.gettransaction(txId)
    return res
  }

  return {
    getResult
  }
}
