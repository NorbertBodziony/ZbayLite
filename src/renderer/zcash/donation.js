export const donationTarget = parseInt(process.env.ZBAY_IS_TESTNET)
  ? 'ztestsapling1e2fr65z5t537cyahpu55fm335a2z26293nvj4wu0jjw55s66r669kaw4k26hmgnk23zeqw2k74n'
  : 'zs1ecsq8thnu84ejvfx2jcfsa6zas2k057n3hrhuy0pahmlvqfwterjaz3h772ldlsgp5r2xwvml9g'
