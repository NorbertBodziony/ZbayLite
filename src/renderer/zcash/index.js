import { remote } from 'electron'
import * as R from 'ramda'
import crypto from 'crypto'

import electronStore from '../../shared/electronStore'
import Zcash from './client'

const envVarOr = (defaultValue, envVar) =>
  R.propOr(defaultValue, envVar)(remote.getGlobal('process').env)

let _client = null
let _username = null
export const credentials = () => {
  const isDev = process.env.NODE_ENV === 'development'
  const username = electronStore.get('username')
  if (isDev || !username) {
    return {
      username: 'zbay',
      password: 'zbay'
    }
  } else {
    return {
      username: username,
      password: electronStore.get('password')
    }
  }
}
export const createClient = () => {
  const rpcCredentials = credentials()
  if (R.isNil(_client) || _username !== rpcCredentials.username) {
    _client = new Zcash({
      url: envVarOr('http://localhost:8334', 'ZBAY_NODE_URL'),
      auth: rpcCredentials
    })
    _username = rpcCredentials.username
  }
  return _client
}
export const createRpcCredentials = () =>
  new Promise((resolve, reject) => {
    const nonceUsername = Math.random().toString()
    const noncePassword = Math.random().toString()
    const password = crypto.createHash('sha256')
    const username = crypto.createHash('sha256')
    password.update(noncePassword)
    username.update(nonceUsername)
    electronStore.set('username', username.digest('hex'))
    electronStore.set('password', password.digest('hex'))
    resolve()
  })
export const getClient = () => {
  const rpcCredentials = credentials()
  if (R.isNil(_client) || _username !== rpcCredentials.username) {
    createClient()
  }
  return _client
}

export default {
  getClient
}
