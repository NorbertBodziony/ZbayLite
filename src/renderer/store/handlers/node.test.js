/* eslint import/first: 0 */
jest.mock('../../vault')
jest.mock('../../zcash', () => ({
  getClient: jest.fn(() => ({
    status: {},
    addresses: {},
    accounting: {}
  }))
}))

import BigNumber from 'bignumber.js'

import create from '../create'
import nodeHandlers from './node'
import nodeSelectors from '../selectors/node'
import { getClient } from '../../zcash'

describe('Node reducer handles', () => {
  const infoMock = jest.fn()
  const createMock = jest.fn()
  const balanceMock = jest.fn()

  let store = null
  beforeEach(() => {
    store = create()
    jest.clearAllMocks()
    getClient.mockImplementation(() => ({
      status: {
        info: infoMock
      },
      addresses: {
        create: createMock
      },
      accounting: {
        balance: balanceMock
      }
    }))
  })

  const assertStoreState = () => expect(
    nodeSelectors.node(store.getState())
  ).toMatchSnapshot()

  it('createAddress', async () => {
    createMock.mockImplementation(async (type) => `${type}-zcash-address`)
    const address = await store.dispatch(nodeHandlers.actions.createAddress('sapling'))
    expect(address).toMatchSnapshot()
  })

  it('rejected createAddress', async () => {
    createMock.mockImplementation(async () => {
      throw Error('createAddress error')
    })
    try {
      await store.dispatch(nodeHandlers.actions.createAddress('sapling'))
    } catch (err) {}
    assertStoreState()
  })

  describe('epics -', () => {
    describe('getStatus', () => {
      it('when successfull', async () => {
        infoMock.mockImplementation(async () => ({
          latestBlock: new BigNumber(2234),
          currentBlock: new BigNumber(12),
          connections: new BigNumber(10),
          isTestnet: true,
          status: 'healthy'
        }))
        await store.dispatch(nodeHandlers.epics.getStatus())
        assertStoreState()
      })

      it('when node is down', async () => {
        infoMock.mockImplementation(async () => {
          throw Error('some kind of node error')
        })
        try {
          await store.dispatch(nodeHandlers.epics.getStatus())
        } catch (err) {}
        assertStoreState()
      })
    })
  })
})
