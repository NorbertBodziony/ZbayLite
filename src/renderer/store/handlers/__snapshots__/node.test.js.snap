// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node reducer handles createAddress 1`] = `
Object {
  "action": Object {
    "payload": "sapling-zcash-address",
    "type": "CREATE_ZCASH_ADDRESS_FULFILLED",
  },
  "value": "sapling-zcash-address",
}
`;

exports[`Node reducer handles epics - getStatus when node is down 1`] = `
Immutable.NodeState {
  "latestBlock": "0",
  "currentBlock": "0",
  "connections": "0",
  "isTestnet": null,
  "status": "down",
  "errors": [Error: some kind of node error],
  "bootstrapLoader": Immutable.Record {
    "loading": false,
    "message": "",
  },
  "startedAt": null,
}
`;

exports[`Node reducer handles epics - getStatus when successfull 1`] = `
Immutable.NodeState {
  "latestBlock": "2234",
  "currentBlock": "12",
  "connections": "10",
  "isTestnet": true,
  "status": "healthy",
  "errors": "",
  "bootstrapLoader": Immutable.Record {
    "loading": false,
    "message": "",
  },
  "startedAt": null,
}
`;

exports[`Node reducer handles rejected createAddress 1`] = `
Immutable.NodeState {
  "latestBlock": "0",
  "currentBlock": "0",
  "connections": "0",
  "isTestnet": null,
  "status": "connecting",
  "errors": [Error: createAddress error],
  "bootstrapLoader": Immutable.Record {
    "loading": false,
    "message": "",
  },
  "startedAt": null,
}
`;
